### PR TITLE
Refactor Embeddings Drivers To Use Chunkers

### PR DESCRIPTION
**Pull request by collindutter**
This PR contains several key changes:

1.  Moved the chunking logic from `OpenAiEmbeddingDriver` into `BaseEmbeddingDriver`. 
2. Refactored Embedding Drivers to use Chunkers when embedding long text. Previously they relied on Tokenizer's `encode` method which not every Tokenizer supported.
3. Fixed Chunkers not chunking contiguous text with no separators. I.e. `foobarfoobar` would not previously be chunked into `foobar` and `foobar`.
4. Removed Tokenizer's `encode` and `decode` methods. Most APIs did not support such functionality, and we no longer need them for embedding long text.

### CODE REVIEW

File: griptape/tokenizers/base_tokenizer.py [hunk 1]
```diff
@@ -27,19 +24,6 @@ def tokens_left(self, text: str) -> int:
...
-        while batch := tuple(islice(it, self.max_tokens)):
-            yield batch
-
     @abstractmethod
-    def encode(self, text: str) -> list[int]:
-        ...
-
-    @abstractmethod
-    def decode(self, tokens: list[int]) -> str:
+    def token_count(self, text: str) -> int:
```
Comment (collindutter) [inline griptape/tokenizers/base_tokenizer.py:28]: While we're here, should this be renamed to `count_tokens`? `token_count` feels more like a property.

File: tests/unit/drivers/embedding/test_azure_openai_embedding_driver.py [hunk 2]
```diff
@@ -18,12 +18,5 @@ def driver(self):
     def test_init(self, driver):
         assert driver
 
-    def test_try_embed_string(self, driver):
-        assert driver.try_embed_string("foobar") == [0, 1, 0]
-
```
Comment (collindutter) [inline tests/unit/drivers/embedding/test_azure_openai_embedding_driver.py:23]: Moved to test for parent class.

File: tests/unit/drivers/embedding/test_azure_openai_embedding_driver.py [hunk 3]
```diff
@@ -18,12 +18,5 @@ def driver(self):
...
     def test_embed_chunk(self, driver):
-        assert driver.embed_chunk("foobar") == [0, 1, 0]
-        assert driver.embed_chunk([1, 2, 3]) == [0, 1, 0]
-
-    def test_embed_long_string(self, driver):
-        assert driver.embed_long_string("foobar" * 5000) == [0, 1, 0]
```
Comment (collindutter) [inline tests/unit/drivers/embedding/test_azure_openai_embedding_driver.py:29]: Moved to test for parent class.

File: tests/unit/drivers/embedding/test_openai_embedding_driver.py [hunk 4]
```diff
@@ -15,30 +15,16 @@ def mock_openai(self, mocker):
     def test_init(self):
         assert OpenAiEmbeddingDriver()
 
-    def test_try_embed_string(self):
-        assert OpenAiEmbeddingDriver().try_embed_string("foobar") == [0, 1, 0]
-
-    def test_try_embed_string_with_long_string(self):
-        assert OpenAiEmbeddingDriver().try_embed_string(
-            " ".join(["foobar"] * 5000)
-        ) == [0, 1, 0]
```
Comment (collindutter) [inline tests/unit/drivers/embedding/test_openai_embedding_driver.py:24]: Moved to test for parent class.

File: tests/unit/loaders/test_pdf_loader.py [hunk 5]
```diff
@@ -22,7 +22,7 @@ def test_load(self, loader):
 
         artifacts = loader.load(path)
 
-        assert len(artifacts) == 149
+        assert len(artifacts) == 151
```
Comment (collindutter) [inline tests/unit/loaders/test_pdf_loader.py:25]: Have not dug into why this changed. I assume it's from some edge-case in the PDF chunking that is now handled by the non-separator fallback method.

Comment (vasinov) [reply tests/unit/loaders/test_pdf_loader.py:25]: Just thought of something...let's add tests for the last chunk that tests the last sentence with `endswith`, so are 100% sure that we are not dropping any chunks in the process.

File: griptape/chunkers/base_chunker.py [hunk 6]
```diff
@@ -48,62 +48,78 @@ def _chunk_recursively(
             tokens_count = 0
             half_token_count = token_count // 2
 
+            # If a separator is provided, only use separators after it.
```
Comment (vasinov) [inline griptape/chunkers/base_chunker.py:51]: Thanks for adding the comments!

File: griptape/drivers/embedding/base_embedding_driver.py [hunk 7]
```diff
@@ -1,21 +1,62 @@
...
 
 
 @define
 class BaseEmbeddingDriver(ExponentialBackoffMixin, ABC):
     dimensions: int = field(kw_only=True)
+    tokenizer: BaseTokenizer = field(kw_only=True)
+    chunker: BaseChunker = field(init=False)
+
+    def __attrs_post_init__(self) -> None:
+        self.chunker = TextChunker(tokenizer=self.tokenizer)
```
Comment (vasinov) [inline griptape/drivers/embedding/base_embedding_driver.py:18]: Let's move this to a `Factory`.

**Review by vasinov — CHANGES_REQUESTED** [review]

File: griptape/drivers/embedding/base_embedding_driver.py (hunk 7 above)
Comment (collindutter) [reply griptape/drivers/embedding/base_embedding_driver.py:18]: Had it as a `Factory` originally but was running into errors about `tokenizer` not being initialized. Let me try again.

Comment (vasinov) [reply griptape/drivers/embedding/base_embedding_driver.py:18]: I think it might have to be `kw_only=True`.

File: griptape/tokenizers/base_tokenizer.py (hunk 1 above)
Comment (collindutter) [reply griptape/tokenizers/base_tokenizer.py]: Done. Also renamed `tokens_left` to `count_tokens_left`

File: tests/unit/loaders/test_pdf_loader.py (hunk 5 above)
Comment (collindutter) [reply tests/unit/loaders/test_pdf_loader.py:25]: Done

File: griptape/drivers/embedding/base_embedding_driver.py (hunk 7 above)
Comment (collindutter) [reply griptape/drivers/embedding/base_embedding_driver.py]: Removed default, child classes must initialize the Chunker.

Comment (collindutter) [reply griptape/drivers/embedding/base_embedding_driver.py:18]: Kept in `__attrs_post_init` based on Slack conversation.

**Review by vasinov — APPROVED** [review]
